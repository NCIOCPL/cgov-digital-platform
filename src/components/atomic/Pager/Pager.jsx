import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Utilities from '../../../utilities/utilities';
import './Pager.scss';

const Pager = ({ data, startFromPage, numberToShow, callback }) => {
  const [currentPage, setCurrentPage] = useState(startFromPage);

  useEffect(() => {
    if (callback) {
      // originally, this assumed that the data comes back as a huge block, in CTS case, each results set is a distinct block
      // const startFrom = currentPage * numberToShow;
      // const endAt = startFrom + numberToShow;
      // const results = data.slice(startFrom, endAt);
      const results = data;
      callback(results, currentPage);
    }
    //document.querySelector('.pager__num--active').focus();
  }, [currentPage, data, numberToShow]);

  useEffect(() => {
    setCurrentPage(startFromPage);
  }, [startFromPage])

  const renderEllipsis = key => {
    return (
      <div key={key} tabIndex="-1" className="pager__num pager__ellipses">
        {'...'}
      </div>
    );
  };

  const determineSteps = (total, numberToShow, activePage) => {
    const divisions = Math.ceil(total / numberToShow);
    const pagesFromStart = activePage;
    const pagesFromEnd = divisions - (activePage + 1);
    let pages;
    if (divisions > 1) {
      if (pagesFromStart > 3) {
        pages = [0, renderEllipsis, activePage - 1, activePage];
      } else {
        pages = Array(activePage + 1)
          .fill()
          .map((el, idx) => idx);
      }
      if (pagesFromEnd > 3) {
        pages = [
          ...pages,
          activePage + 1,
          renderEllipsis,
          divisions - 1,
        ];
      } else {
        const remainingPages = Array(pagesFromEnd)
          .fill()
          .map((el, idx) => activePage + 1 + idx);
        pages = [...pages, ...remainingPages];
      }
    } else {
      //One or fewer divisions exist. Do not render the Pager
      pages = [];
    }
    return pages;
  };

  const determineResults = pageNumber => {
    setCurrentPage(pageNumber);
  };

  const renderSteps = (stepsArray, activePage) => {
    return stepsArray.map((currentStep, idx) => {
      if (typeof currentStep === 'function') {
        return currentStep(idx);
      } else {
        const isCurrent = currentStep.toString() === activePage.toString();
        return (
          <div
            key={idx}
            tabIndex="0"
            className={`pager__num ${isCurrent ? 'pager__num--active' : ''}`}
            onClick={!isCurrent ? () => determineResults(currentStep) : null}
            onKeyPress={Utilities.keyHandler({
              fn: !isCurrent ? () => determineResults(currentStep) : null,
            })}
          >
            {currentStep + 1}
          </div>
        );
      }
    });
  };

  let steps = determineSteps(data.length, numberToShow, currentPage);
  let isFirstPage = currentPage === 0;
  let isLastPage = currentPage + 1 === Math.ceil(data.length / numberToShow);
  return (
    <nav className="pager__container">
      {steps.length > 0 ? (
        <div className="pager__nav">
          {!isFirstPage && (
            <div
              className="pager__arrow"
              tabIndex="0"
              onClick={() => determineResults(currentPage - 1)}
              onKeyPress={Utilities.keyHandler({
                fn: () => determineResults(currentPage - 1),
              })}
            >
              {'< Previous'}
            </div>
          )}
          {renderSteps(steps, currentPage)}
          {!isLastPage && (
            <div
              className="pager__arrow"
              tabIndex="0"
              onClick={() => determineResults(currentPage + 1)}
              onKeyPress={Utilities.keyHandler({
                fn: () => determineResults(currentPage + 1),
              })}
            >
              {'Next >'}
            </div>
          )}
        </div>
      ) : null}
    </nav>
  );
};

Pager.propTypes = {
  data: PropTypes.array,
  startFromPage: PropTypes.number,
  numberToShow: PropTypes.number,
  callback: PropTypes.func.isRequired,
};

Pager.defaultProps = {
  data: [],
  startFromPage: 0,
  numberToShow: 10,
};

export default Pager;
