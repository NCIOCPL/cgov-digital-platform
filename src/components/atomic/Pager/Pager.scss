.pager__container {
  margin-top: 0.6em;
  margin-bottom: 0.6em;

  .pager__count p {
    font-weight: bold;
  }

  .pager__nav {
    margin-left: auto;
    display: flex;
    justify-content: flex-start;
    align-items: center;
  }

  .pager__num,
  .pager__arrow {
    padding: 3px 6px;
    border-radius: 3px;
    font-weight: 400;
    color: #2b7bba;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
    transition: all 150ms linear;
  }

  .pager__num,
  .pager__arrow {
    &:focus {
      background: #cbe8fc;
    }
  }

  .pager__num,
  .pager__arrow {
    margin: 0 3px;
  }

  .pager__num,
  .pager__arrow {
    &:hover {
      background: #cbe8fc;
    }
  }

  .pager__ellipses {
    &:hover,
    &:focus {
      background: white;
      cursor: default;
    }
  }

  .pager__num--active {
    background: #2a71a5;
    color: white;
    cursor: default;
    user-select: none;
    &:hover,
    &:focus {
      cursor: default;
      background: #2a71a5;
    }
  }
}
