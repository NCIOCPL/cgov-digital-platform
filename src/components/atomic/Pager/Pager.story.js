import generateStories from '../../../../.storybook/generateStories';
import Pager from './Pager';

generateStories({
  component: { Pager },
  storyName: 'Elements|Pager',
  defaultProps: {
    data: [...Array(101).keys()],
    numberToShow: 10,
  },
  stories: [
    {
      name: 'default',
    },
  ],
});
