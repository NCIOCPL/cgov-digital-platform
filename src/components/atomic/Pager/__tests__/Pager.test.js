import React from 'react';
import { render, mount } from 'enzyme';
import Pager from '../Pager';

const defaultProps = {
  data: [...Array(55).keys()],
  startFromPage: 0,
  numberToShow: 10,
  callback: jest.fn(),
};

const setup = (enzymeMethod = render, props = {}) => {
  const component = enzymeMethod(<Pager {...props} />);
  return component;
};

describe('Pager', () => {
  describe('Render', () => {
    it('should match a snapshot', () => {
      const component = setup(render, defaultProps);
      expect(component).toMatchSnapshot();
    });
  });

  describe('Logic', () => {
    it('should invoke the callback when a page number is clicked or pressed, returning a slice of the array', () => {
      const component = setup(mount, defaultProps);
      const startFromPageTwo = 1 * defaultProps.numberToShow;
      const slicedArray = defaultProps.data.slice(
        startFromPageTwo,
        startFromPageTwo + defaultProps.numberToShow
      );
      const pages = component.find('.pager__num');
      pages.at(1).simulate('click');
      expect(defaultProps.callback).toHaveBeenCalledWith(slicedArray, 1);
    });
  });
});
