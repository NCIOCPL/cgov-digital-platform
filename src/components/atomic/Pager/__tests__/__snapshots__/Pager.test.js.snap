// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pager Render should match a snapshot 1`] = `
<nav
  class="pager__container"
>
  <div
    class="pager__nav"
  >
    <div
      class="pager__num pager__num--active"
      tabindex="0"
    >
      1
    </div>
    <div
      class="pager__num "
      tabindex="0"
    >
      2
    </div>
    <div
      class="pager__num "
      tabindex="0"
    >
      3
    </div>
    <div
      class="pager__num pager__ellipses"
      tabindex="-1"
    >
      ...
    </div>
    <div
      class="pager__num "
      tabindex="0"
    >
      7
    </div>
    <div
      class="pager__arrow"
      tabindex="0"
    >
      &gt;
    </div>
  </div>
</nav>
`;
